src/java.base/unix/native/libnio/ch/FileDispatcherImpl.c

```diff
@@ -167,9 +169,19 @@ Java_sun_nio_ch_FileDispatcherImpl_force0(JNIEnv *env, jobject this,
 
 #ifdef MACOSX
     result = fcntl(fd, F_FULLFSYNC);
-    if (result == -1 && errno == ENOTSUP) {
-        /* Try fsync() in case F_FULLSYUNC is not implemented on the file system. */
-        result = fsync(fd);
+    if (result == -1) {
+        if (errno == ENOTSUP) {
+            /* Try fsync() in case F_FULLSYUNC is not implemented on the
+             * file system.
+             */
+            result = fsync(fd);
+        } else {
+            struct statfs fbuf;
+            if (fstatfs(fd, &fbuf) == 0 && (fbuf.f_flags & MNT_LOCAL) == 0) {
+                /* Try fsync() in case file is not local. */
+                result = fsync(fd);
+            }
+        }
```

Comment:
Do you need to save/restore errno so that the original error from fcntl(F_FULLSYNC) is used in the error handling?

Also what you would think about dropping the check for ENOTSUP.

Reply:
I thought about `errno` but didn't know whether it's common practice to cache it like that.

We had `ENOTSUP` I suppose in case the `F_FULLSYNC` is unsupported on a local file system. From the man page of `fcntl()` the doc of `F_FULLSYNC` states
```
This is currently implemented on HFS, MS-DOS (FAT), Universal Disk Format (UDF) and APFS file systems
```
although `ENOTSUP` itself is not actually documented here.

Reply:
I think this looks good, I'm just wondering to myself if it is possible to get ENOTSUP from a local device.

Reply:
I am going to try to test this.

Reply:
I was unable to get `ENOTSUP` from local FAT32 or NTFS devices. However from the review discussion [1] of [2] it looks like `ENOTSUP` has been observed and is handled in other packages [3]. Is there any harm to leaving it in?

[1] https://bugs.openjdk.java.net/browse/JDK-8080589
[2] http://mail.openjdk.java.net/pipermail/nio-dev/2015-May/003160.html
[3] https://community.oracle.com/tech/developers/discussion/453477/patch-when-fnctl-fd-f-fullsync-0-returns-enotsup

Reply:
The force method needs  fcntl(F_FULLSYNC) when the file system is local so I think it would be safer to drop the special handle for ENOTSUP. This will mean fallback to fsync when the file is not local.